zap_otlp_encoder/zap_otlp_encoder.go

```diff
@@ -0,0 +1,262 @@
+func (enc *otlpEncoder) AddInt32(k string, v int32)     {}
+func (enc *otlpEncoder) AddInt16(k string, v int16)     {}
+func (enc *otlpEncoder) AddInt8(k string, v int8)       {}
+func (enc *otlpEncoder) AddUint(k string, v uint)       {}
+func (enc *otlpEncoder) AddUint32(k string, v uint32)   {}
+func (enc *otlpEncoder) AddUint16(k string, v uint16)   {}
+func (enc *otlpEncoder) AddUint8(k string, v uint8)     {}
+func (enc *otlpEncoder) AddUintptr(k string, v uintptr) {}
+
+func (enc *otlpEncoder) Clone() zapcore.Encoder {
+	clone := enc.clone()
+	clone.buf.Write(enc.buf.Bytes())
+	return clone
+}
+
+func (enc *otlpEncoder) clone() *otlpEncoder {
+	clone := getOTLPEncoder()
+	clone.EncoderConfig = enc.EncoderConfig
+	clone.spaced = enc.spaced
+	clone.openNamespaces = enc.openNamespaces
+	clone.buf = bufferPool.Get()
+	clone.log = &lpb.LogRecord{}
+	return clone
+}
+
+var levelMap = map[string]uint{
+	"trace": 1,
+	"debug": 5,
+	"info":  9,
+	"warn":  13,
+	"error": 17,
+	"fatal": 21,
+}
+
+func (enc *otlpEncoder) EncodeEntry(ent zapcore.Entry, fields []zapcore.Field) (*buffer.Buffer, error) {
+	final := enc.clone()
+	if final.LevelKey != "" && final.EncodeLevel != nil {
+		final.log.SeverityNumber = lpb.SeverityNumber(levelMap[ent.Level.String()])
+		final.log.SeverityText = ent.Level.String()
+	}
+	if final.TimeKey != "" {
+		final.log.TimeUnixNano = uint64(ent.Time.UnixNano())
+	}
+	if ent.LoggerName != "" && final.NameKey != "" {
+		final.log.Attributes = append(enc.log.Attributes, &v1.KeyValue{
+			Key: "logger",
+			Value: &v1.AnyValue{
+				Value: &v1.AnyValue_StringValue{
+					StringValue: ent.LoggerName,
+				},
+			},
+		})
+	}
+	if ent.Caller.Defined {
+		if final.CallerKey != "" {
+			final.addKeyVal(final.CallerKey, &v1.AnyValue{Value: &v1.AnyValue_StringValue{StringValue: ent.Caller.String()}})
+		}
+		if final.FunctionKey != "" {
+			final.addKeyVal(final.FunctionKey, &v1.AnyValue{Value: &v1.AnyValue_StringValue{StringValue: ent.Caller.Function}})
+		}
+	}
+	if final.MessageKey != "" {
+		final.log.Body = &v1.AnyValue{
+			Value: &v1.AnyValue_StringValue{
+				StringValue: ent.Message,
+			},
+		}
+	}
+
+	addFields(final, fields)
+	data, err := proto.Marshal(final.log)
+	if err != nil {
+		panic(err)
+	}
+
+	final.buf.AppendString(string(data))
+	ret := final.buf
+	putOTLPEncoder(final)
+
+	return ret, nil
+}
+
+func (enc *otlpEncoder) addKeyVal(key string, val *v1.AnyValue) {
+	if key == "trace_id" {
+		traceId, err := trace.TraceIDFromHex(val.GetStringValue())
+		if err == nil {
+			enc.log.TraceId = traceId[:]
+		}
+		return
+	}
+	if key == "span_id" {
+		spanId, err := trace.SpanIDFromHex(val.GetStringValue())
+		if err == nil {
+			enc.log.SpanId = spanId[:]
+		}
+		return
+	}
+
+	enc.log.Attributes = append(enc.log.Attributes, &v1.KeyValue{
+		Key:   key,
+		Value: val,
+	})
+}
+
+func addFields(enc zapcore.ObjectEncoder, fields []zapcore.Field) {
+	for i := range fields {
+		fields[i].AddTo(enc)
+	}
```

Comment:
How are these fields retrieved back? I see syncer is just unmarshalling to logrecord which won't contain the fields?

Reply:
What do you mean by retrieved back ? do you mean how it is sent to otel collector ?

Reply:
Yes, where are they deserialised?

Reply:
It's done by the pushData function in the zap_otlp_sync.

Reply:
Sure let me write a couple of tests.

Reply:
So fields are moved to attributes, if it's something like top-level key of the OTEL log model like traceId, then it is added respectively.

I have added some tests, please check once.